src/rgw/rgw_acl.cc

```diff
@@ -71,27 +77,26 @@ uint32_t RGWAccessControlList::get_group_perm(ACLGroupTypeEnum group,
   return 0;
 }
 
-uint32_t RGWAccessControlList::get_referer_perm(const std::string http_referer,
+uint32_t RGWAccessControlList::get_referer_perm(const uint32_t current_perm,
+                                                const std::string http_referer,
                                                 const uint32_t perm_mask)
 {
   ldout(cct, 5) << "Searching permissions for referer=" << http_referer
                 << " mask=" << perm_mask << dendl;
 
-  /* FIXME: C++11 doesn't have std::rbegin nor std::rend. We would like to
-   * switch when C++14 becomes available. */
-  const auto iter = std::find_if(referer_list.crbegin(), referer_list.crend(),
-    [&http_referer](const ACLReferer& r) -> bool {
-      return r.is_match(http_referer);
+  /* This function is bacically a transformation from current perm to
+   * a new one that takes into consideration the Swift's HTTP referer-
+   * based ACLs. We need to go through all items to respect negative
+   * grants. */
+  uint32_t referer_perm = current_perm;
+  for (const auto& r : referer_list) {
```

Comment:
The case of `.r:-.example.com,.r:*.example.com`:
* RadosGW:
```
rgw$ curl -i "${publicURL}/cont" -X POST -H "X-Auth-Token: ${token}"  -H "X-Container-Read: .r:-.example.com,.r:*.example.com"
HTTP/1.1 204 No Content
X-Trans-Id: tx00000000000000000000b-0058e644b2-1024-default
Content-Type: text/plain; charset=utf-8
Date: Thu, 06 Apr 2017 13:37:54 GMT

rgw$ curl -i "${publicURL}/cont/aaa10" -X GET -H "Referer: https://www.example.com"
HTTP/1.1 200 OK
Content-Length: 3
Accept-Ranges: bytes
Last-Modified: Thu, 06 Apr 2017 13:37:03 GMT
X-Timestamp: 1491485823.30361
etag: 202cb962ac59075b964b07152d234b70
X-Trans-Id: tx00000000000000000000c-0058e64512-1024-default
Content-Type: application/x-www-form-urlencoded
Date: Thu, 06 Apr 2017 13:39:30 GMT

123
```
* Swift:
```
swift$ curl -i "${publicURL}/cont" -X POST -H "X-Auth-Token: ${token}"  -H "X-Container-Read: .r:-.example.com,.r:*.example.com"
HTTP/1.1 204 No Content
Content-Length: 0
Content-Type: text/html; charset=UTF-8
X-Trans-Id: txb9790199c6684ab9931f6-0058e64524
Date: Thu, 06 Apr 2017 13:39:48 GMT

swift$ curl -i "${publicURL}/cont/aaa10" -X GET -H "Referer: https://www.example.com"
HTTP/1.1 200 OK
Content-Length: 3
Accept-Ranges: bytes
Last-Modified: Wed, 25 Jan 2017 18:50:32 GMT
Etag: 202cb962ac59075b964b07152d234b70
X-Timestamp: 1485370231.32458
Content-Type: application/x-www-form-urlencoded
X-Trans-Id: txfe59ab709d004f5ca6acf-0058e6452e
Date: Thu, 06 Apr 2017 13:39:58 GMT

123
```
----

The case of `.r:-.exmple.com,.r:*.example.com` (original, without `a` in the first item):
* RadosGW:
```
rgw$ curl -i "${publicURL}/cont" -X POST -H "X-Auth-Token: ${token}"  -H "X-Container-Read: .r:-.exmple.com,.r:*.example.com"
HTTP/1.1 204 No Content
X-Trans-Id: tx00000000000000000000d-0058e6460e-1024-default
Content-Type: text/plain; charset=utf-8
Date: Thu, 06 Apr 2017 13:43:42 GMT

rgw$ curl -i "${publicURL}/cont/aaa10" -X GET -H "Referer: https://www.example.com"
HTTP/1.1 200 OK
Content-Length: 3
Accept-Ranges: bytes
Last-Modified: Thu, 06 Apr 2017 13:37:03 GMT
X-Timestamp: 1491485823.30361
etag: 202cb962ac59075b964b07152d234b70
X-Trans-Id: tx00000000000000000000e-0058e64616-1024-default
Content-Type: application/x-www-form-urlencoded
Date: Thu, 06 Apr 2017 13:43:50 GMT

123
```
* Swift:
```
swift$ curl -i "${publicURL}/cont" -X POST -H "X-Auth-Token: ${token}"  -H "X-Container-Read: .r:-.exmple.com,.r:*.example.com"
HTTP/1.1 204 No Content
Content-Length: 0
Content-Type: text/html; charset=UTF-8
X-Trans-Id: txf89be419de33481191fea-0058e64637
Date: Thu, 06 Apr 2017 13:44:23 GMT

swift$ curl -i "${publicURL}/cont/aaa10" -X GET -H "Referer: https://www.example.com"
HTTP/1.1 200 OK
Content-Length: 3
Accept-Ranges: bytes
Last-Modified: Wed, 25 Jan 2017 18:50:32 GMT
Etag: 202cb962ac59075b964b07152d234b70
X-Timestamp: 1485370231.32458
Content-Type: application/x-www-form-urlencoded
X-Trans-Id: tx3ea6d1f8b6fe4e4fbdbfd-0058e6463a
Date: Thu, 06 Apr 2017 13:44:26 GMT

123
```